### ISSUE TITLE
Decompression fails on second message with javascript client

### ISSUE THREAD

**Issue by bridger** [IBM-Swift/Kitura-WebSocket-NIO#22]
I'm seeing a decompression assertion failure when the Javascript client connects and sends the second binary message.

It is hitting this assert
https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/e53219402a26f8448a309c0b0be59f2c72c95d1f/Sources/KituraWebSocket/PermessageDeflateDecompressor.swift#L136 
with rc=-3

It works if I disable the permessage-deflate extension here:
https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/beceef8db142f1c1f19a405cbeaccf5fde05b74b/Sources/KituraWebSocket/WSConnectionUpgradeFactory.swift#L31

As a side-note, an API to disable permessage-deflate might be useful. For example, there may be a client with a bad implementation. It might be nice to inspect the incoming request and decide which extensions should be supported.

**Comment by bridger** [comment]
I have a guess for what might be causing this bug. In this codebase the zlib stream is initialized for every new message: https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/e53219402a26f8448a309c0b0be59f2c72c95d1f/Sources/KituraWebSocket/PermessageDeflateDecompressor.swift#L90

In the SwiftWebsocket project the zlib stream is only initialized once per connection. The same stream is reused each frame. Compression would be more efficient if it can use information from previous frames.

**Comment by bridger** [comment]
Here is where SwiftWebsocket initializes the z_stream. It is wrapped in a Swift class:
https://github.com/tidwall/SwiftWebSocket/blob/1a8b805ab75053e7387a3dbfaacc2bc11b1c47de/Source/WebSocket.swift#L418

That Inflater class is created once when the connection opens if permessage-deflate is enabled: https://github.com/tidwall/SwiftWebSocket/blob/1a8b805ab75053e7387a3dbfaacc2bc11b1c47de/Source/WebSocket.swift#L1183

**Comment by pushkarnk** [comment]
Yes, that sounds right. We initialise a new zlib stream on every new message if the client does NOT send `server_no_context_takeover` (which means the the context is taken over) as a part of its negotiation handshake. We must initialise the zlib stream only once if this extension parameter is not received. I think that is the case here. 

[This line](https://github.com/tidwall/SwiftWebSocket/blob/1a8b805ab75053e7387a3dbfaacc2bc11b1c47de/Source/WebSocket.swift#L1011) indicates that if the property `noContextTakeover` is not set to true, `server_no_context_takeover` will not be sent, meaning we will create the zlib stream only once.

**Comment by pushkarnk** [comment]
@bridger can you please confirm that you are NOT setting `compression.noContextTakeover` in your tests?

**Comment by bridger** [comment]
@pushkarnk In the headers of the connection request I don't see anything for `server_no_context_takeover` or `client_no_context_takeover`.

I see that the PermessageDeflateDecompressor and PermessageDeflateCompressor are each initialized with noContextTakeOver = false. However, those variables are only ever written to - they aren't used in the code after initialization it looks like.

In this case I'm using Firefox or Chrome as the websocket client. They don't have any APIs exposed to control whether `server_no_context_takeover` or `client_no_context_takeover` are used. There isn't even an API to disable the permessage-deflate extension or modify the headers in any way. The Javascript API is very limited.

The spec is hard to read, but it sounds like the server should send back `client_no_context_takeover` or else it will be assumed it is using takeover: https://tools.ietf.org/html/rfc7692#section-7.1.1.2

>    Absence of this extension parameter in an extension negotiation
>    response indicates that the server can decompress messages built by
>    the client using context takeover.

**Comment by bridger** [comment]
I worked on this bug a little more today. I was able to get Firefox connected and using compression!

It looks like when the client sends `server_no_context_takeover` or `client_no_context_takeover` it is just a suggestion for the server. It is up to the server to send back those parameters if it decides to do no_context_takeover.

Because the code always does a no_context_takeover today, we should always send those parameters back (until context takeover is supported).

I put in a hack so that `func negotiate(header: String) -> String` always returns `permessage-deflate; client_no_context_takeover; server_no_context_takeover` and it worked!

We'd need a less hacky way to send back the response. Ideally it would be based on PermessageDeflateCompressor.noContextTakeOver and PermessageDeflateDecompressor.noContextTakeover, but those objects aren't available in `func negotiate(header: String) -> String`

**Comment by pushkarnk** [comment]
@bridger Glad that you found the workaround. Can you share instructions and/or a test case to have this issue reproduced? Whatever literature I've come across vaguely says the "compression context is reused". I think we need to first write a valid test case that represents this scenario.

I understand that when the compression context is shared across multiple messages, which means we don't initialise (or even reset) the zlib stream on every frame receipt. However, if I simply try reusing (i.e initialising it only on the first message and then reusing it without resetting), I see the same precondition failure (possible because `stream.next_in` is nil).

**Comment by pushkarnk** [comment]
@bridger I'll work on defaulting to "no context takeover" so that even if a client doesn't request for it (wants context takeover), we'll deny. Later, once we have context takeover supported, we'll remove this default.

**Comment by pushkarnk** [comment]
Though the specification is very loosely defined (or must we say it is flexible!) about context takeover, to keep things simple I propose we have this in the longer run (after we've figured out context takeover):

- Client sends **client_no_context_takover** in a negotiation offer
   - include  **client_no_context_takover** in the negotiation response
   - do not use context takeover for decompressing messages received from the client i.e initialise a new zlib stream for decompressing every incoming message)

- Client DOES NOT send **client_no_context_takover** in a negotiation offer
   - DO NOT include  **client_no_context_takover** in the negotiation response
   - Use context takeover for decompressing incoming messages(reuse the zlib stream for decompressing incoming messages)

- Client sends **server_no_context_takover** in a negotiation offer
   - Include **server_no_context_takover** in the negotiation response
   - do not use context takeover for compressing outgoing messages (use a new zlib stream every time)

- Client does not **server_no_context_takover** in a negotiation offer
  - Do not include **server_no_context_takover** in the negotiation response
  - Use context takeover to compress  outgoing messages (reuse the zlib stream for every message)

**Comment by pushkarnk** [comment]
Before we figure out an implementation for context take-over: we can have the work-around suggested [here](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/issues/22#issuecomment-491599830). Irrespective of what the client includes in its negotiation offer, we'll send `client_no_context_takeover; server_no_context_takeover` in the response.

**Comment by bridger** [comment]
> Can you share instructions and/or a test case to have this issue reproduced?

Sure, you can repro this by running a bit of Javascript in Firefox or Chrome. You can just paste this in the browser's console:

```
  let ws = this.ws = new WebSocket("localhost:5000/ws");
  ws.onopen = (event) => {
    console.log("Websocket opened");

    ws.send("This is the first message that was sent");
    ws.send("This is the second message that was sent");

  };
```

I like your plan for the workaround. Thank you for putting it into a pull request!

It makes sense to have the server default to using context takeover, unless the client requests no context takeover. That plan would work well for my application. Other authors using Kitura Websocket might want more control over the memory usage of their server - maybe there should be an option on the server side to force no-context-takeover to always be on?

**Comment by pushkarnk** [comment]
some new issues found with permessage-deflate and two relevant findings:

https://github.com/IBM-Swift/Kitura-WebSocket-NIO/issues/32#issuecomment-503436793

https://github.com/IBM-Swift/Kitura-WebSocket-NIO/issues/32#issuecomment-503461572

**Comment by pushkarnk** [comment]
Closing based on feedback on #26

**Comment by pushkarnk** [comment]
Fixes available in release 2.1.0
